# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/ #broken link
****

In [2]:
import pandas as pd

## imports for Python, Pandas

In [3]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [4]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [7]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [8]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population info.governor    state shortname
0        Dade       12345    Rick Scott  Florida        FL
1     Broward       40000    Rick Scott  Florida        FL
2  Palm Beach       60000    Rick Scott  Florida        FL
3      Summit        1234   John Kasich     Ohio        OH
4    Cuyahoga        1337   John Kasich     Ohio        OH

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [9]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

[{u'_id': {u'$oid': u'52b213b38594d8a2be17c780'},
  u'approvalfy': 1999,
  u'board_approval_month': u'November',
  u'boardapprovaldate': u'2013-11-12T00:00:00Z',
  u'borrower': u'FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA',
  u'closingdate': u'2018-07-07T00:00:00Z',
  u'country_namecode': u'Federal Democratic Republic of Ethiopia!$!ET',
  u'countrycode': u'ET',
  u'countryname': u'Federal Democratic Republic of Ethiopia',
  u'countryshortname': u'Ethiopia',
  u'docty': u'Project Information Document,Indigenous Peoples Plan,Project Information Document',
  u'envassesmentcategorycode': u'C',
  u'grantamt': 0,
  u'ibrdcommamt': 0,
  u'id': u'P129828',
  u'idacommamt': 130000000,
  u'impagency': u'MINISTRY OF EDUCATION',
  u'lendinginstr': u'Investment Project Financing',
  u'lendinginstrtype': u'IN',
  u'lendprojectcost': 550000000,
  u'majorsector_percent': [{u'Name': u'Education', u'Percent': 46},
   {u'Name': u'Education', u'Percent': 26},
   {u'Name': u'Public Administration, Law, and Ju

In [13]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')
#sample_json_df

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

# My submissions

Here, I load in the file and get some basic info.

In [4]:
projects = pd.read_json('data/world_bank_projects.json')

projects.columns

Index(['_id', 'approvalfy', 'board_approval_month', 'boardapprovaldate',
       'borrower', 'closingdate', 'country_namecode', 'countrycode',
       'countryname', 'countryshortname', 'docty', 'envassesmentcategorycode',
       'grantamt', 'ibrdcommamt', 'id', 'idacommamt', 'impagency',
       'lendinginstr', 'lendinginstrtype', 'lendprojectcost',
       'majorsector_percent', 'mjsector_namecode', 'mjtheme',
       'mjtheme_namecode', 'mjthemecode', 'prodline', 'prodlinetext',
       'productlinetype', 'project_abstract', 'project_name', 'projectdocs',
       'projectfinancialtype', 'projectstatusdisplay', 'regionname', 'sector',
       'sector1', 'sector2', 'sector3', 'sector4', 'sector_namecode',
       'sectorcode', 'source', 'status', 'supplementprojectflg', 'theme1',
       'theme_namecode', 'themecode', 'totalamt', 'totalcommamt', 'url'],
      dtype='object')

There appears to be many columns that aren't relevant for this exercise, so I decided to reshape the data by picking out a few important columns.

In [6]:
projects = projects[['_id', 'approvalfy', 'country_namecode', 'countrycode', 'countryname', 'countryshortname', 'mjtheme', 'mjtheme_namecode', 'project_name']]

projects.shape

(500, 9)

In [8]:
#Examine the first few rows of the new dataframe
projects.head(4)

_id  approvalfy  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014   

                               country_namecode countrycode  \
0  Federal Democratic Republic of Ethiopia!$!ET          ET   
1                      Republic of Tunisia!$!TN          TN   
2                                   Tuvalu!$!TV          TV   
3                        Republic of Yemen!$!RY          RY   

                               countryname    countryshortname  \
0  Federal Democratic Republic of Ethiopia            Ethiopia   
1                      Republic of Tunisia             Tunisia   
2                                   Tuvalu              Tuvalu   
3                        Republic of Yemen  Yemen, Republic of   

                                             mjtheme  \
0                                [Human development]   
1  [Economic management, Social protection and ri...   
2  [Trade and integration, Public sector governan...   
3  [Social dev/gender/inclusion, Social dev/gende...   

                                    mjtheme_namecode  \
0  [{'code': '8', 'name': 'Human development'}, {...   
1  [{'code': '1', 'name': 'Economic management'},...   
2  [{'code': '5', 'name': 'Trade and integration'...   
3  [{'code': '7', 'name': 'Social dev/gender/incl...   

                                        project_name  
0  Ethiopia General Education Quality Improvement...  
1          TN: DTF Social Protection Reforms Support  
2  Tuvalu Aviation Investment Project - Additiona...  
3   Gov't and Civil Society Organization Partnership

Now we're ready to dive in.

## Question 1: Top 10 countries with most projects

For this exercise, I grouped the data by country and counted the number of projects for each one. I then ordered the result and extracted the top ten.

In [9]:
grouped = projects[['countryshortname', 'project_name']].groupby('countryshortname').count()

#Now order the result in descending order
ordered = grouped.sort_values('project_name', ascending=False)

#Then, extract the first 10 rows from the ordered set and print the result
top_10 = ordered.iloc[:10, :]

top_10

project_name
countryshortname                
China                         19
Indonesia                     19
Vietnam                       17
India                         16
Yemen, Republic of            13
Nepal                         12
Bangladesh                    12
Morocco                       12
Mozambique                    11
Africa                        11

One curious thing to point out is that Africa is listed as a country name, even though we can clearly see that individual African countries are alson on the list. If we explore further, we can see that the data contains info on regional projects which don't apply to one specific country. 

If we wanted, we could remove those rows from the data to only count projects for specific countries.

In [11]:
#Investigating rows for Africa
projects[projects['countryshortname'] == 'Africa'].head(3)

_id  approvalfy country_namecode  \
45  {'$oid': '52b213b38594d8a2be17c7ad'}        2014      Africa!$!3A   
46  {'$oid': '52b213b38594d8a2be17c7ae'}        2014      Africa!$!3A   
51  {'$oid': '52b213b38594d8a2be17c7b3'}        2014      Africa!$!3A   

   countrycode countryname countryshortname  \
45          3A      Africa           Africa   
46          3A      Africa           Africa   
51          3A      Africa           Africa   

                                              mjtheme  \
45                                [Human development]   
46  [Public sector governance, Rural development, ...   
51  [Environment and natural resources management,...   

                                     mjtheme_namecode  \
45  [{'code': '8', 'name': 'Human development'}, {...   
46  [{'code': '2', 'name': 'Public sector governan...   
51  [{'code': '11', 'name': 'Environment and natur...   

                                         project_name  
45  West Africa  Regional Disease Surveillance Cap...  
46  RCIP4 - Regional Communications Infrastructure...  
51  Building Resilience through Innovation, Commun...

In [12]:
#recalculated top-10 list without Africa
remove_africa = projects[projects['countryshortname'] != 'Africa']

regrouped = remove_africa[['countryshortname', 'project_name']].groupby('countryshortname').count()

reordered = regrouped.sort_values('project_name', ascending=False)

new_top_10 = reordered.iloc[:10, :]

new_top_10

project_name
countryshortname                
China                         19
Indonesia                     19
Vietnam                       17
India                         16
Yemen, Republic of            13
Morocco                       12
Nepal                         12
Bangladesh                    12
Mozambique                    11
Burkina Faso                   9

## Question 2: Top 10 major project themes

The column of focus for this exercise is the one labelled 'mjtheme_namecode', where each entry is a list of themes that describe each project. Each theme in the list is a dictionary broken down as a numerical code and the name that applies to each number. 

In [14]:
#Investigate the column of interest
themes = projects['mjtheme_namecode']

themes.head(3)

0    [{'code': '8', 'name': 'Human development'}, {...
1    [{'code': '1', 'name': 'Economic management'},...
2    [{'code': '5', 'name': 'Trade and integration'...
Name: mjtheme_namecode, dtype: object

In [15]:
#Inspect one entry to see the full list for that row
themes[1]

[{'code': '1', 'name': 'Economic management'},
 {'code': '6', 'name': 'Social protection and risk management'}]

With the column extracted, my first approach was to count the frequency of each code:

In [16]:
#Set up the dict that will keep track of the counts
counter = {}

#Now iterate through the codes in each row and add them to the dictionary
for row in themes:
    for d in row:
        if d['code'] in counter:
            counter[d['code']] += 1
        else:
            counter[d['code']] = 1
            
counter

{'8': 210,
 '11': 250,
 '1': 38,
 '6': 168,
 '5': 77,
 '2': 199,
 '7': 130,
 '4': 146,
 '10': 216,
 '9': 50,
 '3': 15}

This provides a count for each code, but the dictionary is unordered.

A more efficient way would be to use a counter.

In [17]:
#Import the necessary module
import collections

#Initiate an empty Counter, and then update it after iterating through the themes series
count = collections.Counter()

for row in themes:
    for d in row:
        count[d['code']] += 1
        
#Now get the top 10 by using the most_common method
count.most_common(10)

[('11', 250),
 ('10', 216),
 ('8', 210),
 ('2', 199),
 ('6', 168),
 ('4', 146),
 ('7', 130),
 ('5', 77),
 ('9', 50),
 ('1', 38)]

This allows us to clearly see which codes are use most frequently. 

However, in keeping with the theme of the assignment, I though it might be better to try to solve this problem using json_normalize.

In [18]:
#Load in the data file
data = json.load(open('data/world_bank_projects.json'))

#Now normalize the 'mjtheme_namecode' column and inspect the results
themes_table = json_normalize(data, 'mjtheme_namecode', ['id'])

themes_table.head(3)

code                 name       id
0    8    Human development  P129828
1   11                       P129828
2    1  Economic management  P144674

Now I did another groupby operation, this time grouping by the name column:

In [19]:
#Groupby the name, sort in descending order, extract the top 10 entries
themes_table[['name', 'code']].groupby('name').count().sort_values('code', ascending=False).head(10)

code
name                                              
Environment and natural resources management   223
Rural development                              202
Human development                              197
Public sector governance                       184
Social protection and risk management          158
Financial and private sector development       130
                                               122
Social dev/gender/inclusion                    119
Trade and integration                           72
Urban development                               47

The above gives us names of each theme and their frequency in the data.

However, the results don't match the previous counter method, which had a top result of 250 for code 11. That's because there are 122 instances where the name was left blank.

Because of this, I decided to also group by code on the normalized data:

In [21]:
themes_table[['name', 'code']].groupby('code').count().sort_values('name', ascending=False).head(10)

name
code      
11     250
10     216
8      210
2      199
6      168
4      146
7      130
5       77
9       50
1       38

These results match the results from before

## Question 3: Fill in the missing names

For this task, my approach was to create a dictionary of all the unique codes and their corresponding names, and then reference that dictionary to fill in any missing values.

In [23]:
#First, create the dictionary

#get of all the codes and all the names that occur in the themes table 
#and zip them together into a combined list
codes = list(themes_table['code'])
names = list(themes_table['name'])
zipped = list(zip(codes, names))

#Now redefine the zipped variable as a sorted set of unique values
zipped = sorted(list(set(zipped)))

zipped

[('1', ''),
 ('1', 'Economic management'),
 ('10', ''),
 ('10', 'Rural development'),
 ('11', ''),
 ('11', 'Environment and natural resources management'),
 ('2', ''),
 ('2', 'Public sector governance'),
 ('3', ''),
 ('3', 'Rule of law'),
 ('4', ''),
 ('4', 'Financial and private sector development'),
 ('5', ''),
 ('5', 'Trade and integration'),
 ('6', ''),
 ('6', 'Social protection and risk management'),
 ('7', ''),
 ('7', 'Social dev/gender/inclusion'),
 ('8', ''),
 ('8', 'Human development'),
 ('9', ''),
 ('9', 'Urban development')]

In [26]:
#Now remove all the pairs with empty strings, resulting a list of the correct combinations
for tup in zipped:
    if tup[1] == '':
        zipped.remove(tup)
    
zipped

[('1', 'Economic management'),
 ('10', 'Rural development'),
 ('11', 'Environment and natural resources management'),
 ('2', 'Public sector governance'),
 ('3', 'Rule of law'),
 ('4', 'Financial and private sector development'),
 ('5', 'Trade and integration'),
 ('6', 'Social protection and risk management'),
 ('7', 'Social dev/gender/inclusion'),
 ('8', 'Human development'),
 ('9', 'Urban development')]

In [27]:
#Next, convert the list into a dict in order to reference the key:value pairs
d = dict(zipped)

In [28]:
#Now, using the themes Series from the first part of question 2, update the 
#series by referencing dict, d, and replacing missing values
for entry in themes:
    for theme in entry:
        if theme['name'] == '':
            theme['name'] = d[theme['code']]

Finally, with the themes Series updated, I used that to update the original projects data set by replacing the flawed 'mjtheme_namecode' column:

In [29]:
projects['mjtheme_namecode'] = themes